o2cb_ctl/o2cb_ctl.8.in
.TH "o2cb_ctl" "8" "September 2010" "Version @VERSION@" "OCFS2 Manual Pages"
.SH "NAME"
o2cb_ctl \- Control program for the \fIO2CB\fR cluster service.
.SH "SYNOPSIS"
.B o2cb_ctl 
\fB\-C\fR \fB\-n\fR \fIobject\fR \fB\-t\fR type\fR [\fB\-i\fR] [\fB\-a\fR \fIattribute\fR ]
.br
.B o2cb_ctl 
\fB\-D\fR \fB\-n\fR \fIobject\fR [\fB\-u\fR]
.br
.B o2cb_ctl 
\fB\-I\fR [\fB\-o\|\-z\fR] \fB\-l\fR \fImanager\fR> [\fB\-n\fR \fIobject\fR>] [\fB\-t\fR \fItype\fR] [\fB\-a\fR \fIattribute\fR]
.br
.B o2cb_ctl 
\fB\-H\fR [\fB\-n\fR \fIobject\fR] [\fB\-t\fR \fItype\fR>] [\fB\-a\fR \fIattribute\fR>] 
.br
.B o2cb_ctl 
\fB\-h\fR
.br
.B o2cb_ctl 
\fB\-V\fR

.SH "DESCRIPTION"
.PP 
\fBo2cb_ctl\fR is the control program for the \fIO2CB\fR cluster service.
Users are not advised to use this program directly but instead use the
\fIO2CB\fR init service and/or \fIocfs2console\fR.

.SH "OPTIONS"

.TP
\fB\-C\fR
Create an object in the OCFS2 Cluster Configuration.

.TP
\fB\-D\fR
Delete an object from the existing OCFS2 Cluster Configuration.

.TP
\fB\-I\fR
Print information about the OCFS2 Cluster Configuration.

.TP
\fB\-H\fR
Change an object or objects in the existing OCFS2 Cluster Configuration.

.TP
\fB\-h\fR
Displays help and exit.

.TP
\fB\-V\fR
Print version and exit.

.SH "OTHER OPTIONS"

.TP
\fB\-a\fR \fI<attribute>\fR
With \fB\-C\fR, \fI<attribute>\fR is in format "parameter=value", where
the parameter is a valid parameter that can be set in the file
\fI/etc/ocfs2/cluster.conf\fR.  With \fB\-I\fR, \fI<attribute>\fR may be
"parameter", indicating an attribute to be listed in the output, or it
may be "parameter==value", indicating that only objects matching
"parameter=value" are to be displayed.

.TP
\fB\-i\fR
Valid only with \fB\-C\fR. When creating something (\fInode\fR or
\fIcluster\fR), it will also install it in the live cluster. If the
parameter is not specified, then only update the
\fI/etc/ocfs2/cluster.conf\fR.

.TP
\fB\-n\fR \fIobject\fR
\fIobject\fR is usually the node name or cluster name. In the
\fI/etc/ocfs2/cluster.conf\fR file, it would be the value of the name
parameter for any of the sections (\fIcluster\fR or \fInode\fR).

.TP
\fB\-o\fR
Valid only with \fB\-I\fR. Using this parameter, if one asks o2cb_ctl to
list all nodes, it will output it in a format suitable for shell parsing.

.TP
\fB\-t\fR \fItype\fR
\fItype\fR can be cluster, node or heartbeat.

.TP
\fB\-u\fR
Valid only with \fB\-D\fR. When deleting something (\fInode\fR or
\fIcluster\fR), it will also remove it from the live cluster. If the
parameter is not specified, then only update the
\fI/etc/ocfs2/cluster.conf\fR.

.TP
\fB\-z\fR
Valid only with \fB\-I\fR. This is the default. If one asks o2cb_ctl to
list all nodes, it will give a verbose listing.

.SH "EXAMPLES"

.HP
Add node5 to an offline cluster:

$ o2cb_ctl -C -n node5 -t node -a number=5
.br
	-a ip_address=192.168.0.5 -a ip_port=7777
.br
	-a cluster=mycluster

.TP
Add node10 to an online cluster:

$ o2cb_ctl -C -i -n node10 -t node -a number=10
.br
	-a ip_address=192.168.1.10 -a ip_port=7777
.br
	-a cluster=mycluster

Note the -i argument.

.TP
Query the IP address of node5:

$ o2cb_ctl -I -n node5 -a ip_address

.TP
Change the IP address of node5:

$ o2cb_ctl -H -n node5 -a ip_address=192.168.1.5

.SH "SEE ALSO"
.BR mkfs.ocfs2(8)
.BR fsck.ocfs2(8)
.BR tunefs.ocfs2(8)
.BR mounted.ocfs2(8)
.BR o2cb(7)

.SH "AUTHORS"
Oracle Corporation

.SH "COPYRIGHT"
Copyright \(co 2004, 2010 Oracle. All rights reserved.


debian/NEWS
ocfs2-tools (1.2.1-1) unstable; urgency=low

  * The init script needs modules now in order to function properly.  Upstream
    strongly discourages the changes I did in 1.1.5 in order to work for the
    monolithic kernel case.
    The sctp module is no longer necessary.  Please read
    /usr/share/doc/ocfs2-tools/README.Debian for the complete requirements.

 -- David Martínez Moreno <ender@debian.org>  Sun, 28 May 2006 13:18:50 +0200


ocfs2console/ocfs2interface/o2cb_ctl.py
# OCFS2Console - GUI frontend for OCFS2 management and debugging
# Copyright (C) 2002, 2005 Oracle.  All rights reserved.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 021110-1307, USA.

from cStringIO import StringIO

from process import Process
from guiutil import error_box

DEFAULT_CLUSTER_NAME = 'ocfs2'

O2CB_INIT = '/etc/init.d/o2cb'
O2CB_CTL = 'o2cb_ctl'

class CtlError(Exception):
    pass

class O2CBProcess(Process):
    def __init__(self, args, desc, parent=None):
        if isinstance(args, basestring):
            command = '%s %s' % (self.program, args)
        else:
            command = (self.program,) + tuple(args)

        Process.__init__(self, command, self.title, desc, parent)

class O2CBCtl(O2CBProcess):
    program = O2CB_CTL
    title = 'Cluster Control'

class O2CBInit(O2CBProcess):
    program = O2CB_INIT
    title = 'Cluster Stack'

def init_load(parent=None):
    args = ('load',)
    o2cb_init = O2CBInit(args, 'Starting cluster stack...', parent)
    return o2cb_init.reap()

def init_online(cluster_name, parent=None):
    desc = 'Starting %s cluster...' % cluster_name
    args = ('online', cluster_name)
    o2cb_init = O2CBInit(args, desc, parent)
    return o2cb_init.reap()

def init_status(cluster_name=None, parent=None):
    if cluster_name: 
        desc = 'Querying status of %s...' % cluster_name
        args = ('status', cluster_name)
    else:
        desc = 'Querying cluster software status...'
        args = ('status',)
    o2cb_init = O2CBInit(args, desc, parent)
    return o2cb_init.reap()

def query_clusters(parent=None):
    args = '-I -t cluster -o'
    o2cb_ctl = O2CBCtl(args, 'Querying cluster...', parent)
    return o2cb_ctl.reap()

def query_nodes(parent=None):
    args = '-I -t node -o'
    o2cb_ctl = O2CBCtl(args, 'Querying nodes...', parent)
    return o2cb_ctl.reap()

def add_node(name, cluster_name, ip_address, ip_port, parent=None):
    desc = 'Adding node %s...' % name
    args = ('-C', '-n', name, '-t', 'node',
            '-a', 'cluster=%s' % cluster_name,
            '-a', 'ip_address=%s' % ip_address,
            '-a', 'ip_port=%s' % ip_port,
            '-i')

    o2cb_ctl = O2CBCtl(args, desc, parent)
    return o2cb_ctl.reap()

def get_active_cluster_name(parent=None):
    cluster_name = None

    success, output, k = query_clusters(parent)

    if success:
        names = []

        buffer = StringIO(output)

        for line in buffer:
            if line.startswith('#'):
                continue

            try:
                name, rest = line.split(':', 1)
            except ValueError:
                 continue

            names.append(name)

        if DEFAULT_CLUSTER_NAME in names:
            cluster_name = DEFAULT_CLUSTER_NAME
        elif len(names):
            cluster_name = names[0]

    if cluster_name is None:
        args = '-C -n %s -t cluster -i' % DEFAULT_CLUSTER_NAME
        o2cb_ctl = O2CBCtl(args, 'Creating cluster...', parent)
        success, output, k = o2cb_ctl.reap()

        if success:
            cluster_name = DEFAULT_CLUSTER_NAME
        else:
            errmsg = '%s\nCould not create cluster %s' % (output,
                                                          DEFAULT_CLUSTER_NAME)
            raise CtlError, errmsg

    return cluster_name

def get_cluster_nodes(cluster_name, parent=None):
    success, output, k = query_nodes(parent)
    if not success:
        raise CtlError, output

    nodes = []

    buffer = StringIO(output)

    for line in buffer:
        if line.startswith('#'):
            continue

        try:
            name, cluster, node, ip_address, ip_port, state = line.split(':')
        except ValueError:
            continue

        if cluster == cluster_name:
            info = {}
            symtab = locals()

            for sym in 'name', 'node', 'ip_address', 'ip_port':
                info[sym] = symtab[sym]

            nodes.append(info)

    return nodes

def main():
    success, output, k = init_load()
    if success:
        print 'Success:\n' + output
    else:
        print 'Failed:\n' + output

    success, output, k = query_nodes()
    if success:
        print 'Success:\n' + output
    else:
        print 'Failed:\n' + output

if __name__ == '__main__':
    main()
